File: apps/server/src/routes/meta.ts
import { Elysia, t } from 'elysia'
import { requireAuth } from '$src/auth/gate'
import { getConnection } from '$src/store/connections'
import { getDriver } from '$src/db/registry'

export const metaRoutes = new Elysia({ prefix: '/connections' })
  .use(requireAuth)
  // Reject unknown connection ids early for every route in this group.
  .resolve(async ({ params, status }) => {
    const id = (params as { id?: string }).id
    if (!id || !getConnection(id)) return status(404, { error: 'Connection not found' })
    return { driver: await getDriver(id) }
  })
  .get('/:id/databases', ({ driver }) => driver.introspect.listDatabases())
  .get('/:id/schemas', ({ driver }) => driver.introspect.listSchemas())
  .get('/:id/tree', ({ driver, query }) => driver.introspect.getTree(query.search), {
    query: t.Object({ search: t.Optional(t.String()) }),
  })
  .get(
    '/:id/tables/:schema/:table',
    async ({ driver, params }) => {
      const [columns, indexes, constraints, primaryKey] = await Promise.all([
        driver.introspect.getColumns(params.schema, params.table),
        driver.introspect.getIndexes(params.schema, params.table),
        driver.introspect.getConstraints(params.schema, params.table),
        driver.introspect.getPrimaryKey(params.schema, params.table),
      ])
      return { columns, indexes, constraints, primaryKey }
    },
    { params: t.Object({ id: t.String(), schema: t.String(), table: t.String() }) },
  )
  // Entities for the editor completion service. Tables + columns are the core of
  // intellisense; functions + FKs are enrichments. The enrichments must never be
  // able to break core completion, so a failure there degrades to an empty list
  // (e.g. pg_proc.prokind needs PG >= 11) instead of failing the whole request.
  .get('/:id/completion', async ({ driver }) => {
    const [tree, columns] = await Promise.all([
      driver.introspect.getTree(),
      driver.introspect.getAllColumns(),
    ])
    const [functions, foreignKeys] = await Promise.all([
      driver.introspect.getFunctions().catch((e) => {
        console.warn('completion: getFunctions failed, skipping', e)
        return []
      }),
      driver.introspect.getForeignKeys().catch((e) => {
        console.warn('completion: getForeignKeys failed, skipping', e)
        return []
      }),
    ])
    return {
      tables: tree.flatMap((s) =>
        s.relations.map((r) => ({ schema: s.schema, name: r.name, type: r.type })),
      ),
      columns,
      functions,
      foreignKeys,
    }
  })

File: apps/server/src/routes/meta.rs
use std::sync::Arc;
use axum::{
    extract::{Path, Query, State},
    response::Json,
    routing::get,
    Router,
};

pub use geto_core::services::meta::TreeQuery;
use geto_core::db::types::{
    CompletionResponse, DatabaseInfo, SchemaTree, TableDetailResponse,
};
use crate::error::AppError;
use crate::state::AppState;

pub fn router() -> Router<Arc<AppState>> {
    Router::new()
        .route("/connections/{id}/databases", get(list_databases_handler))
        .route("/connections/{id}/schemas", get(list_schemas_handler))
        .route("/connections/{id}/tree", get(get_tree_handler))
        .route(
            "/connections/{id}/tables/{schema}/{table}",
            get(get_table_detail_handler),
        )
        .route("/connections/{id}/completion", get(get_completion_handler))
}

#[utoipa::path(
    get,
    path = "/api/connections/{id}/databases",
    params(
        ("id" = String, Path, description = "Connection ID")
    ),
    responses(
        (status = 200, description = "List databases in connection", body = Vec<DatabaseInfo>),
        (status = 404, description = "Connection not found")
    )
)]
pub async fn list_databases_handler(
    State(state): State<Arc<AppState>>,
    Path(id): Path<String>,
) -> Result<Json<Vec<DatabaseInfo>>, AppError> {
    let databases = geto_core::services::meta::list_databases(&state.registry, &state.core, &id).await?;
    Ok(Json(databases))
}

#[utoipa::path(
    get,
    path = "/api/connections/{id}/schemas",
    params(
        ("id" = String, Path, description = "Connection ID")
    ),
    responses(
        (status = 200, description = "List schemas in connection", body = Vec<String>),
        (status = 404, description = "Connection not found")
    )
)]
pub async fn list_schemas_handler(
    State(state): State<Arc<AppState>>,
    Path(id): Path<String>,
) -> Result<Json<Vec<String>>, AppError> {
    let schemas = geto_core::services::meta::list_schemas(&state.registry, &state.core, &id).await?;
    Ok(Json(schemas))
}

#[utoipa::path(
    get,
    path = "/api/connections/{id}/tree",
    params(
        ("id" = String, Path, description = "Connection ID"),
        TreeQuery
    ),
    responses(
        (status = 200, description = "Get database schema tree", body = Vec<SchemaTree>),
        (status = 404, description = "Connection not found")
    )
)]
pub async fn get_tree_handler(
    State(state): State<Arc<AppState>>,
    Path(id): Path<String>,
    Query(q): Query<TreeQuery>,
) -> Result<Json<Vec<SchemaTree>>, AppError> {
    let tree = geto_core::services::meta::get_tree(&state.registry, &state.core, &id, q.search.as_deref()).await?;
    Ok(Json(tree))
}

#[utoipa::path(
    get,
    path = "/api/connections/{id}/tables/{schema}/{table}",
    params(
        ("id" = String, Path, description = "Connection ID"),
        ("schema" = String, Path, description = "Schema name"),
        ("table" = String, Path, description = "Table name")
    ),
    responses(
        (status = 200, description = "Get table details (columns, indexes, constraints)", body = TableDetailResponse),
        (status = 404, description = "Table or connection not found")
    )
)]
pub async fn get_table_detail_handler(
    State(state): State<Arc<AppState>>,
    Path((id, schema, table)): Path<(String, String, String)>,
) -> Result<Json<TableDetailResponse>, AppError> {
    let detail = geto_core::services::meta::get_table_detail(&state.registry, &state.core, &id, &schema, &table).await?;
    Ok(Json(detail))
}

#[utoipa::path(
    get,
    path = "/api/connections/{id}/completion",
    params(
        ("id" = String, Path, description = "Connection ID")
    ),
    responses(
        (status = 200, description = "Get auto-completion metadata", body = CompletionResponse),
        (status = 404, description = "Connection not found")
    )
)]
pub async fn get_completion_handler(
    State(state): State<Arc<AppState>>,
    Path(id): Path<String>,
) -> Result<Json<CompletionResponse>, AppError> {
    let completion = geto_core::services::meta::get_completion(&state.registry, &state.core, &id).await?;
    Ok(Json(completion))
}
